src/pydsl/scope.py

```diff
@@ -2,30 +2,63 @@
 import typing
 from collections.abc import Iterator
 from contextlib import contextmanager
+from functools import reduce
 from typing import Any
 
 
+class BindingContext:
+    """
+    A context manager that temporarily binds a name to a value in a given
```

Comment:
context manager has a specific meaning in python. Since you never run `with BindingContext(...) as b:`, this isn't a context manager.